File: LeetCode/1025. Divisor Game/Solution2.py
class Solution:
    def divisorGame(self, n: int) -> bool:
        # Time and Space Complexity: O(N (1 + 1 / 2 + ... + 1 /N)) = O(N log N)
        divisors = {}

        for divisor in range(1, n + 1):
            i = 2

            while divisor * i <= n:
                num = divisor * i

                if num not in divisors:
                    divisors[num] = set()

                divisors[num].add(divisor)
                i += 1

        self.divisors = divisors
        self.dp = {}

        return self.play(n)

    def play(self, n):
        if n in self.dp:
            return self.dp[n]

        if n == 1:
            return False

        ret = False
        for divisor in self.divisors[n]:
            ret = ret or not (self.play(n - divisor))

        self.dp[n] = ret
        return ret

File: LeetCode/2400. Number of Ways to Reach a Position After Exactly k Steps/Solution2.cpp
class Solution {
private:
    const int MOD = 1000000007;
    const int INV_POW = MOD - 2;
    unordered_map<int, int> factorial_memo;

    // x^(-1) mod MOD == x^(MOD - 2) mod MOD
    long inverse(int value) {
        long ret = 1;
        long power = value;
        long exponent = INV_POW;

        while (exponent > 0) {
            if ((exponent & 1) > 0)
                ret = ret * power % MOD;

            exponent >>= 1;
            power = power * power % MOD;
        }

        return ret;
    }


    long factorial(int n) {
        if (n <= 1)
            return 1;

        if (factorial_memo.find(n) == factorial_memo.end())
            factorial_memo[n] = factorial(n - 1) * n % MOD;

        return factorial_memo[n];
    }
public:
    int numberOfWays(int startPos, int endPos, int k) {
        // Time Complexity O(K + log(MOD))
        // Space Complexity: O(K)

        int dist = abs(endPos - startPos);

        if (dist > k || (dist + k) % 2 == 1)
            return 0;

        int num_right = (k + dist) / 2;
        int num_left = (k - dist) / 2;
        // cout << num_right << " " << num_left << "\n";

        return factorial(k) * inverse(factorial(num_right) * factorial(num_left) % MOD) % MOD;
     }
};
